User defined column remapping

senior opened the pull request:
This PR adds support for 3 new kinds of projections, following the similar pattern/form of the existing arithmetic expressions.

SQLExpressions - Allows arbitrary SQL to be named and projected out as part of the query

RemapExpressions - Allow enum-like values to be remapped to descriptive strings and projected out separately

FKRemapExpressions - Same as RemapExpressions, but gets the textual representation from another table using an FK relationship

***
Remaining back-end work

- [x] Make slack and email pulses aware of remapping
- [x] Revert 95a65d4, query_metadata code should call that for field values
- [x] Add a new enum special type
- [x] Reproduce and fix the protocol missing issue Maz found
- [x] Add logic to clear dimensions if base_type/special_type changes to something that doesn't allow dimensions
- [x] Add sorting support for external remapping
- [ ] Circle back to pulse and add some tests
- [x] Recheck security of new dimension/remapping routes, backfill field values tests

iethree committed:
Add a test helper function for test scaffolding cleanup

This is of most use at the REPL when you want to invoke a
function (such as expectations/run-tests) and have the test
scaffolding available and cleanup after itself.

iethree committed:
Add support for remapping user provided enum-like values

This commit adds storage of dimension data that the user provides for
an existing field. This is to translate enum-like
fields (i.e. integers) to a textual description of that data. This
remapped value will be included in the response with the original
value. The UI will then take that and show the more user friendly
remapped value.

The new "remapped_from" and "remapped_to" attributes in the cols field
of the response will indicate that a field replaces another field.

senior commented on src/metabase/query_processor/expand.clj:
```diff
@@ -393,21 +393,47 @@
+   mapping :- {s/Str s/Str}]
+  (i/map->RemapExpression {:column column
+                           :mapping (reduce-kv (fn [acc k v]
+                                                 (assoc acc (Integer/parseInt k) v))
```
This assumes that the data coming in is integers (which I'm encoding as a string after a discussion with @camsaul). Although it seems like integers are definitely the most common usage of this, I've also seen single character values used as a type of enum as well. Are integers a good place to start here? Or should we make the interchange here support any valid type?

salsakran replied:
What are the implications of not assuming they're integers? This seems like the kind of thing where we're eventually going to need to support arbitrary remapping, and constraining it to integers is probably going to result in more work overall. 

Is there something really weird about letting arbitrary variable types get remapped?

senior commented on src/metabase/driver/generic_sql/query_processor.clj:
```diff
@@ -77,6 +77,22 @@
 
+  SQLExpression
+  (formatted [{:keys [expr]}]
+    (hsql/raw expr))
```
I've been thinking about this SQLExpression stuff. I think this might be expanding the surface area of our API by quite a bit. Would probably be worth talking through some of the implications of this.

senior commented on src/metabase/query_processor/expand.clj:
```diff
@@ -393,21 +393,47 @@
+  [sql-expr :- su/NonBlankString]
+  (i/map->SQLExpression {:expr sql-expr}))
+
+(s/defn ^:ql remap-expression
```
@tlrobinson suggested `remap-inline` for this and `remap-lookup` instead of `fk-remap-expression`. I'm a :+1: on better names for this.

senior commented:
I tested this using the Reviews table from the sample dataset. This would normally be setup by the admin via the UI, but to simulate that locally, I inserted a new row in the metabase_dimensions table
```
insert into metabase_dimensions (field_id, name, type,human_readable_field_id)
VALUES (33, 'Rating Desc', 'internal', null);
```
and then updated the related field values to have human_readable_values
```
update metabase_fieldvalues
set human_readable_values='["Horrible","Bad","Meh","Good","Excellent"]'
where id=1;
```

Then I asked a question about the Reviews table and I see an additional column in the UI named "Rating Desc" with the human_readable_values above.

iethree committed:
Split expand and resolve into separate middleware

This commit splits the what was the expand-resolve middleware separate
expand and resolve middlewares. This also splits the source table
resolving to a separate middleware. This split sets up being able to
add an FK field after query expansion to be resolved normally as it
goes through the pipeline.

tlrobinson committed:
MVP frontend implementation of displaying remapped columns

senior commented on test/metabase/query_processor/middleware/add_dimension_projections_test.clj:
```diff
@@ -0,0 +1,181 @@
+  (-> example-query
+      ;; This needs to be fixed, we don't know the name of the new
+      ;; column until it gets resolved, so it doesn't get set here
+      #_(update-in [:query :fields 2] assoc :remapped-to "TITLE" :remapped-from nil)
```
@salsakran @camsaul @tlrobinson 

The PR currently works for FK remapping by letting the foreign key get resolved like it would normally and that defines the name of the remapped field. Is that the right field name for this? For internal remapping we let the user decide the name of the field. Should the user also be able to decide the name of the FK remapped field?

If so, should that name be used just as the display_name of the column? Or as the name and display_name? Not sure if this plays into the answer or not, but remapped-to/remapped-from are two new fields that @tlrobinson and I added to indicate a source/destination field in a result set. Currently that field uses the name of the source/destination of the remapping.

senior replied:
I've fixed this up based on the following:

- remapped_to/remapped_from always point to a column by its "name" attribute
- FK remapped columns keep their original FK based names (i.e. "TITTLE" from PRODUCT)
- FK remapped columns include the user defined names as the "field_display_name"

Talking with @tlrobinson about it, it seems like this is the best approach.

senior commented:
Pushed up some commits for handling FK related remapping. To test this out I followed the example from the comment above and added another row to the dimensions table:

```
insert into metabase_dimensions (field_id,name,type,human_readable_field_id)
values (32,'Product','external',27)
```

tlrobinson commented:
For those using H2

```shell
# review.product_id => product.title
lein h2 -sql "insert into metabase_dimensions (field_id,name,type,human_readable_field_id) values (32,'Product','external',27)"
# order.product_id => product.title
lein h2 -sql "insert into metabase_dimensions (field_id,name,type,human_readable_field_id) values (3,'Product','external',27)"
# order.user_id => people.name
lein h2 -sql "insert into metabase_dimensions (field_id,name,type,human_readable_field_id) values (7,'User','external',16)"
```

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3646,3 +3646,51 @@ databaseChangeLog:
+      author: senior
+      changes:
+        - createTable:
+            tableName: metabase_dimensions
```
This table name really doesn't need a `metabase_` prefix. The only reason some of our tables have that is for historic reasons -- the original version of Metabase was developed using Django which adds a prefix to DB table names depending on which module they belong to (this is also why some of our tables are prefixed with `report_`)

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3646,3 +3646,51 @@ databaseChangeLog:
+      changes:
+        - createTable:
+            tableName: metabase_dimensions
+            remarks: ''
```
I would definitely add remarks to the new tables and columns explaining their purpose,
so potential contributors or other people trying to hack Metabase don’t ask you a million questions about why the stuff doesn’t do what they expected it to do.

Otherwise you'll have to explain that the purpose of the table is something completely different than what they assumed. Just speaking from experience, a little bit of extra dox will save you GH conversations down the road

camsaul commented:
The table shouldn't be prefixed with `metabase_`. It's obvious that it's for Metabase since it's in the Metabase DB

camsaul commented on test/metabase/query_processor/expand_resolve_test.clj:
```diff
@@ -153,12 +177,18 @@
+                                            :remapped-from      nil
+                                            :remapped-to        nil
+                                            :field-display-name nil}
+                              :value       {:field-placeholder {:field-id           (id :users :last_login)
```
Instead of having to update a million fields it's probably better to do something like 

```clojure
(def ^:private field-placeholder-defaults
  {:remapped-from      nil
   :remapped-to        nil
   :field-display-name nil})

...

{:field-placeholder (merge field-placeholder-defaults 
                           {...}}
```

that way when new things are added to the fields in the future you only need to update one thing instead of dozens of tests.

(We are following a similar pattern in other places in the code -- search for `card-defaults` for example)

iethree committed:
Refactor field test values to reduce duplication

This commit just defs a map with a default map of key/values for
fields/placeholders/columns and merges that with a map of what's
different for that test.

Fixes up eastwood warnings, add docstrings to public functions

tlrobinson committed:
Port Object detail to a visualization and mode

iethree committed:
Add dimensions/human_readable_values to the query_metadata response

This moves values to be nested under the related fields object, and
includes dimension along with human_readable_values in the field as
well.

senior commented:
@tlrobinson Added the query metadata as discussed. The values key is now under fields, it's a list of 1 or 2 item lists ([value] or [value, human_readable_value) along with dimensions.

I changed the name of the table from metabase_dimensions to dimensions as well, so you'll need to run that SQL manually if you've got a database from the previous commits in this PR.

iethree committed:
Minor code cleanup and refactors for remapping

iethree committed:
Added auditing columns to the dimensions table

tlrobinson committed:
Show human readable field values and remapped column name everywhere on frontend

iethree committed:
Add FK related remapping

This commit adds logic around an FK remapping found in the
dimensions table. It will follow a foreign key to the related
table and use the indicated column from that table as the remapped
value for the source column.

This commit also switches the "remapped_from" and "remapped_to"
attributes returned in the query response to use the "name" of the
remapped column as there could be ambiguity with the previously used
display name.

iethree committed:
Add internal remapping support to breakouts

tlrobinson committed:
Fix frontend lint + test errors

iethree committed:
java.jdbc column renames should update metadata

Previously when java.jdbc renamed a project column (i.e. NAME to
NAME_2) to disambiguate them, we would discard previous information on
that NAME column and reconstruct it based on the data that comes back
from the query. This changes that behavior to update the existing
column data to the new name. When we do not have metadata on that
column it will use the existing behavior to infer that.

This commit also adds some tests for remapping which uncovered this
bug.

senior commented:
I removed the WIP on this PR and I think it's ready for review. There are currently two open issues/TODO items. First there is no admin side to this yet as that work still needs to be done (both frontend and backend). There's also a known bug around breakout columns and FK based remapping, that will be fixed up later.

Since the admin work hasn't been done yet, to manually test this you'll need to run some SQL. For "internal" remapping, where the admin defines what values should be remapped to, you'll need to add human readable values and a row in the new dimensions table. Below is an example of an internal remapping for our ratings table to give a description to the 1-5 ratings:

```
insert into dimensions (field_id, name, type,human_readable_field_id, created_at, updated_at)
VALUES (33, 'Rating Desc', 'internal', null, current_timestamp, current_timestamp);

update metabase_fieldvalues
set human_readable_values='["Horrible","Bad","Meh","Good","Excellent"]'
where id=1;
```
A FK based remapping (or "external" remapping) needs to have a row in the dimensions table with the relevant field id that should be used as the display value:

```
insert into dimensions (field_id,name,type,human_readable_field_id, created_at, updated_at)
values (32,'Product','external',27, current_timestamp, current_timestamp);
```
This will use the foreign key associated with field id 32. It will follow that foreign key and use field id 27 from that other table as the remapped value. This example swaps the Product ID with the name of the product.

@tlrobinson gave an example inserting data into h2 with lein above. The table name has changed from metabase_dimensions to dimensions and I've added auditing columns, below is a new copy of the example:

``` 
# review.product_id => product.title
lein h2 -sql "insert into dimensions (field_id,name,type,human_readable_field_id,created_at,updated_at) values (32,'Product','external',27,current_timestamp,current_timestamp)"
# order.product_id => product.title
lein h2 -sql "insert into dimensions (field_id,name,type,human_readable_field_id,created_at,updated_at) values (3,'Product','external',27,current_timstamp,current_timestamp)"
# order.user_id => people.name
lein h2 -sql "insert into dimensions (field_id,name,type,human_readable_field_id,created_at,updated_at) values (7,'User','external',16,current_timestamp,current_timestamp)"

The UI currently will replace the original column with it's remapped value, using the name the admin specified (this is the name column in the dimensions table).

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3646,3 +3646,67 @@ databaseChangeLog:
+                    deferrable: false
+                    foreignKeyName: fk_dimension_displayfk_ref_field_id
+                    initiallyDeferred: false
+                    nullable: true
```
you have two `nullable: true`s here 🙀

camsaul commented:
I read through all the backend code and it looks good to me. You should make sure to run the tests against all drivers by including `[ci drivers]` in one of your commit messages. You could do that by just rewording the last commit and force-pushing.

It will be interesting trying to fix the merge conflicts between this and my `nested-queries` branch 😱

iethree committed:
Added realiasing code for Oracle column remapping

This commit fixes an issue with Oracle when using remapping, with a
column name that is ambiguous and a limit. Those queries are wrapped
in another query, which will fail in Oracle due to the ambiguous
column name. Columns will not be re-aliased if they collide in Oracle
to prevent this issue. Most of this code was written by
@camsaul (Thanks Cam!).

iethree committed:
Transform remapped values to the correct type before comparison

In some databases (i.e. Oracle) query responses can contain different
datatypes than what the JSON can support (the encoding for the values
column of the field values table). This commit adds a funciton that
transforms the values from the field values table to the format the
database will return.

iethree committed:
Fixed up remapping issues on MongoDB

Specifically added code that matches projections from mongo with the
expected order of results coming back from the database. Also added
code to ensure order by columns are compared correctly after the split
of expand and resolve into two separate middlewares.

iethree committed:
Added admin routes for dimensions and human readable values

This commit adds support for adding dimensions to existing
fields. Values and human readable values (optional) can also be
added/updated.

iethree committed:
Druid fixes for remappings [ci drivers]

Tests were failing when attempting to reconcile expected columns with
the actual columns returned. This commit adds code to create a list of
expected columns as the druid query is built.

kdoh committed:
initial field settings page

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3670,3 +3670,68 @@ databaseChangeLog:
+      author: senior
+      changes:
+        - createTable:
+            tableName: dimensions
```
All of our other table names are singulars of the entity that they store attributes of. e.g. `pulse` instead of `pulses`

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -3670,3 +3670,68 @@ databaseChangeLog:
+                    initiallyDeferred: false
+                    nullable: false
+                    references: metabase_field(id)
+                    deleteCascade: true
```
not sure if you need it or not but just a heads up if you want an index on your FKs you'll have to add it manually

senior replied:
This is a good catch, it should have an index

camsaul commented on src/metabase/api/field.clj:
```diff
@@ -71,24 +72,70 @@
+  [id :as {{dimension-type :type dimension-name :name human_readable_field_id :human_readable_field_id} :body}]
+  {dimension-type         (s/enum "internal" "external")
+   dimension-name         su/NonBlankString
+   human_readable_field_id (s/maybe s/Int)}
```
You can also use `su/IntGreaterThanZero` here because a negative or zero ID wouldn't make sense. Not a big difference but heads up that it exists

camsaul commented on src/metabase/api/field.clj:
```diff
@@ -71,24 +72,70 @@
+                     :human_readable_values (when human-readable-values?
+                                              (map second value-pairs))))))
+
+(defn- create-field-values
```
Since function has side effects I would vote it end in a bang, e.g. `create-field-values!`. Calling it twice would not be the same as calling it once; in other words, it wouldn't be safe inside an STM transaction. https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark

camsaul commented on src/metabase/api/field.clj:
```diff
@@ -71,24 +72,70 @@
 
-;; TODO - not sure this is used anymore
-(api/defendpoint POST "/:id/value_map_update"
+(defn- update-field-values [field-value-id value-pairs]
```
same thing, https://github.com/bbatsov/clojure-style-guide#changing-state-fns-with-exclamation-mark

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -79,6 +80,14 @@
+                   (if (seq values)
+                     (update field :values fv/field-values->pairs)
+                     field))
+                 fields))))
```
I think a `for` loop would be a little more readable here, e.g.

```clojure
(defn- format-fields-for-response [resp]
  (update resp :fields 
          (fn [fields]
            (for [{:keys [values] :as field} fields]
              (if (seq values)
                (update field :values fv/field-values->pairs)
                field)))))
```

Generally I think it's better to use higher-level constructs like `for` over `map` where sensible

senior replied:
I typically use for when I need filtering or some of the other list comprehension features vs. avoiding an anonymous fn, but I don't have strong opinions on it, I don't mind switching it.

camsaul replied:
it doesn't really make a huge difference whichever one you think is more readable

camsaul commented on src/metabase/driver/googleanalytics/query_processor.clj:
```diff
@@ -4,6 +4,7 @@
             [clojure.tools.reader.edn :as edn]
             [medley.core :as m]
             [metabase.query-processor.util :as qputil]
+            [metabase.query-processor.middleware.expand :as ql]
```
nitpicky but these requires are out-of-order. You can do `M-x cljr-clean-ns` to sort it automatically for you

senior replied:
I'm doing this manually right now, an emacs fn would be much better. Did you get that function here? https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-clean-ns

camsaul commented on src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -197,11 +197,14 @@
+      pipeline-ctx
+      (let [projections (doall (map #(vector (->lvalue %) (->initial-rvalue %)) all-fields))]
+        (-> pipeline-ctx
+            (assoc  :projections (doall (map (comp keyword first) projections)))
```
can't you just use `mapv` instead of `doall + map`?

senior replied:
I had doall in there to debug an issue, it should be removed. Good catch!

camsaul commented on test/metabase/api/table_test.clj:
```diff
@@ -504,3 +501,49 @@
   []
   ((user->client :crowberto) :get 200 "table/card__1000/fks"))
+
+(defn narrow-fields [category-names api-response]
```
if this isn't being used outside this namespace I would vote that we make it private. That way if we need to tweak it in the future we know it only affects tests in this namespace so we don't need to grep the codebase looking for other usages

camsaul commented on test/metabase/driver/bigquery_test.clj:
```diff
@@ -3,12 +3,15 @@
              [util :as tu]]
             [metabase.test.data.datasets :refer [expect-with-engine]]))
 
+(def col-defaults
```
same thing, if we make this `^:private` it will make our lives easier should we choose to refactor it in the future

camsaul commented on test/metabase/driver/druid_test.clj:
```diff
@@ -35,19 +35,24 @@
                              :database (data/id)})
           (m/dissoc-in [:data :results_metadata])))))
 
+(def ^:private col-defaults
```
`^:private` = 👍

camsaul commented on test/metabase/driver/druid_test.clj:
```diff
@@ -35,19 +35,24 @@
-                             {:name "venue_price", :display_name "Venue Price", :base_type :type/Text}
-                             {:name "venue_name",  :display_name "Venue Name",  :base_type :type/Text}
-                             {:name "count",       :display_name "Count",       :base_type :type/Integer}]
+
```
extra newline

camsaul commented on test/metabase/driver/mongo_test.clj:
```diff
@@ -78,7 +78,9 @@
    :data      {:rows        [[1]]
                :columns     ["count"]
-               :cols        [{:name "count", :display_name "Count", :base_type :type/Integer}]
+
```
extra newline

camsaul commented on test/metabase/query_processor/expand_resolve_test.clj:
```diff
@@ -19,6 +23,36 @@
+(def ^:private resolve'
+  (comp resolve/resolve (st/resolve-source-table-middleware identity)))
+
+(def field-ph-defaults
```
`^:private`

camsaul commented on test/metabase/query_processor/expand_resolve_test.clj:
```diff
@@ -19,6 +23,36 @@
+   :remapped-to        nil
+   :field-display-name nil})
+
+(def field-defaults
```
`^:private`

camsaul commented on test/metabase/query_processor/expand_resolve_test.clj:
```diff
@@ -19,6 +23,36 @@
+   :dimensions      []
+   :values          []})
+
+(def price-field-values
```
`^:private`

https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

camsaul commented on test/metabase/query_processor/expand_resolve_test.clj:
```diff
@@ -19,6 +23,36 @@
                                {k (obj->map v)}))
     :else           o))
 
+(def ^:private resolve'
```
Even by reading the code for this function I'm not really sure what's it's supposed to. It's not really clear from the name (`resolve'`) so it would be helpful to me or anyone else who has to work on it in the future to have either a more descriptive name or a short docstring or even just a comment explaining the purpose

camsaul commented on test/metabase/query_processor/middleware/add_dimension_projections_test.clj:
```diff
@@ -0,0 +1,185 @@
+
+(tu/resolve-private-vars metabase.query-processor.middleware.add-dimension-projections remap-results add-fk-remaps)
+
+(def col-defaults
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

camsaul commented on test/metabase/query_processor/middleware/add_dimension_projections_test.clj:
```diff
@@ -0,0 +1,185 @@
+     :remapped_from nil,
+     :remapped_to nil})
+
+(def example-resultset
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

camsaul commented on test/metabase/query_processor/middleware/add_dimension_projections_test.clj:
```diff
@@ -0,0 +1,185 @@
+                        :remapped_to nil}))))
+  (remap-results example-resultset))
+
+(def field-defaults
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

camsaul commented on test/metabase/query_processor/middleware/add_dimension_projections_test.clj:
```diff
@@ -0,0 +1,185 @@
+   :values [],
+   :visibility-type :normal})
+
+(def example-query
```
https://github.com/metabase/metabase/wiki/Metabase-Clojure-Style-Guide#make-things-private

camsaul commented on test/metabase/query_processor/middleware/add_dimension_projections_test.clj:
```diff
@@ -0,0 +1,185 @@
+                                            :field-display-name "Product"}))
+  (add-fk-remaps example-query))
+
+(def external-remapped-result
```
`^:private`

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -285,3 +287,50 @@
   `(with-temp-db [_# (resolve-dbdef '~dataset)]
      ~@body))
+
+(defn delete-model-instance!
```
`defn-`

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -285,3 +287,50 @@
+  [{:keys [id] :as instance}]
+  (db/delete! (-> instance name symbol) :id id))
+
+(defn call-with-data
```
`defn-`

senior replied:
This one is only used by the `with-data` macro, but when the macro expands, clojure complains about `call-with-data` being private, so I'll leave this as public.

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -285,3 +287,50 @@
+        (doseq [instance result-instances]
+          (delete-model-instance! instance))))))
+
+(defmacro with-data [data-load-fn & body]
```
`defmacro ^:private`

senior replied:
This one is used outside of the ns

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -285,3 +287,50 @@
+(defmacro with-data [data-load-fn & body]
+  `(call-with-data ~data-load-fn (fn [] ~@body)))
+
+(def  venue-categories
```
`^:private`

senior replied:
This one is used outside of the ns

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -285,3 +287,50 @@
+(def  venue-categories
+  (map vector (defs/field-values defs/test-data-map "categories" "name")))
+
+(defn create-venue-category-remapping
```
`defn-`

senior replied:
This one is only used outside of this ns

camsaul commented on test/metabase/test/data.clj:
```diff
@@ -285,3 +287,50 @@
+                              :values (json/generate-string (range 0 (count venue-categories)))
+                              :human_readable_values (json/generate-string (map first venue-categories))})]))
+
+(defn create-venue-category-fk-remapping
```
`defn-`

senior replied:
This one is only used outside of this ns

camsaul replied:
oops sorry 😓

camsaul commented on test/metabase/test/data/dataset_definitions.clj:
```diff
@@ -34,3 +34,20 @@
 ;; sender and reciever -- allowing us to test situations where multiple joins for a *single* table should occur.
 (def-database-definition-edn avian-singles)
+
+(def test-data-map
```
`^:private`

senior replied:
This one is only used outside of this namespace, but used in conjunction with the field values function below

camsaul replied:
ok carry on

camsaul commented on test/metabase/test/data/dataset_definitions.clj:
```diff
@@ -34,3 +34,20 @@
+                       (zipmap field-names row)))))
+          {} (:table-definitions test-data)))
+
+(defn field-values
```
`defn-`

senior replied:
This one is only used outside of this namespace

camsaul commented on test/metabase/test_setup.clj:
```diff
@@ -99,3 +99,17 @@
   (log/info "Shutting down Metabase unit test runner")
   (core/stop-jetty!))
+
+(defn call-with-test-scaffolding
```
Is this used anywhere? Or if it's only for REPL usage can we also make it `^:private`?

senior replied:
It's only used at the REPL so should be fine to mark as private, I can just invoke it via the var

camsaul replied:
in that case I guess it doesn't really make a big difference, you can leave it public

camsaul reviewed (changes requested):
Ok, this looks good to me except for a handful of style suggestions. 95% of them are things that we should probably make `^:private` to make our lives easier in the future. But fundamentally the rest of it looks good

senior commented:
@camsaul thanks for the feedback. I was thinking the private vars was a src thing, not a tests thing, I'll get them fixed up.

kdoh committed:
disable cache section for now

camsaul commented:
The private vars don't matter as much in tests, but I think it should still make things a little easier to work with in the long run 👍  thanks

iethree committed:
Rename the dimensions table to dimension

iethree committed:
Minor code style changes [ci drivers]

Mostly this commit marks test functions as private if their not used
outside of their ns

senior commented:
@camsaul Thanks for the feedback! I fixed up the things you mentioned, currently waiting for the full test run.

kdoh committed:
add button

attekei committed:
Add metadata loading & back button logic

iethree committed:
Remapping admin fixes for the backend

Added a route for deleting dimensions, and added field_id to the field
values response.

attekei committed:
Working custom mapping editor

attekei committed:
Save the chosen mapping type to `dimension` table

iethree committed:
Remove the category type predicate from remapping

The logic predates remapping, but since remapping used human readable
values, the predicates were kept in place. This commit changes that to
generate field values if it is not present and the user has asked for it.

attekei committed:
First (not very successful) take on foreign key picker

attekei committed:
Working FK selection, only support integer custom mappings

attekei committed:
Restrict Foreign Key mapping to only FK fields

attekei committed:
Hide field type and field visiblity temporarily

kdoh commented:
Couple things I found while playing with the user facing side of things in the QB. All of these are based off of remapping the `sample dataset` -> `review` ->`rating` column to custom values.

1. Looks like the X axis on a bar chart is getting a bit confused and is trying to fit the remapped names onto a numerical axis which doesn't make much sense.
![screen shot 2017-07-10 at 12 01 11 pm](https://user-images.githubusercontent.com/5248953/28035042-003eec0a-6568-11e7-95d1-64b631a4f282.png)

2. Looks like selected filters aren't being properly reflected in the filter drop down.
![screen shot 2017-07-10 at 12 02 08 pm](https://user-images.githubusercontent.com/5248953/28035143-5e2b359e-6568-11e7-885e-919ead97d7bc.png)

kdoh committed:
add placeholders

kdoh committed:
layout tweaks

kdoh committed:
fix column settings layout + hover state

attekei committed:
Add visibility and field type pickers

attekei committed:
Add stubs for integrated tests

attekei committed:
Fix lint errors

iethree committed:
Merge remote-tracking branch 'upstream/master' into custom-user-projections [ci drivers]

attekei committed:
Integrated tests WIP

attekei committed:
Add frontend tests for almost all admin stuff

attekei committed:
Merge branch 'master' of https://github.com/metabase/metabase into custom-user-projections

attekei committed:
Add tests for custom remappings [ci e2e]

attekei committed:
Lint fix [ci e2e]

attekei commented:
My first bug findings:

- FK remappings don't work if the target field is a date. When trying to map Product ID field of Orders to the creation date of Product, I see the Product Id instead of creation date in results.

- In table settings, you see both original and remapped field but and they are both enabled. You can't change which fields are shown (the checkboxes are no-op, although you can save the settings). See the duplicate fields here:
<img width="388" alt="screen shot 2017-07-11 at 12 07 13" src="https://user-images.githubusercontent.com/6034928/28085756-89fd8810-6631-11e7-8221-3f8f0224709b.png">

- If you have set a wrong target field for a foreign key special type, and also have set a FK remapping for that field, "Raw Data" queries (and probably other queries which use that field as well?) result in a confusing error. Example where I tried to use User ID as a target for Reviewer field (although Reviewer actually isn't a foreign key):

```
JdbcSQLException:
 Message: Data conversion error converting "vidal.moore"; SQL statement:
-- Metabase:: userID: 1 queryType: MBQL queryHash: f222558cb024da4b784de79776a1eeca81fce972281c8527d6683d8c67c01707
SELECT "PUBLIC"."REVIEWS"."ID" AS "ID", "PUBLIC"."REVIEWS"."BODY" AS "BODY", "PUBLIC"."REVIEWS"."CREATED_AT" AS "CREATED_AT", "PUBLIC"."REVIEWS"."PRODUCT_ID" AS "PRODUCT_ID", "PUBLIC"."REVIEWS"."RATING" AS "RATING", "PUBLIC"."REVIEWS"."REVIEWER" AS "REVIEWER", "PRODUCTS__via__PRODUCT_ID"."TITLE" AS "TITLE", "PEOPLE__via__REVIEWER"."NAME" AS "NAME" FROM "PUBLIC"."REVIEWS" LEFT JOIN "PUBLIC"."PRODUCTS" "PRODUCTS__via__PRODUCT_ID" ON "PUBLIC"."REVIEWS"."PRODUCT_ID" = "PRODUCTS__via__PRODUCT_ID"."ID" LEFT JOIN "PUBLIC"."PEOPLE" "PEOPLE__via__REVIEWER" ON "PUBLIC"."REVIEWS"."REVIEWER" = "PEOPLE__via__REVIEWER"."ID" LIMIT 2000 [22018-194]
 SQLState: 22018
 Error Code: 22018
```
Hopefully we could show a more helpful message (like "Your foreign key might be incorrectly set up"). We talked with @senior that in general we should try to produce good error messages

attekei committed:
Filter dates out, choose entity name as default in FK remapping

attekei committed:
Fix the filter widget showing undefined for normal field values

attekei committed:
Fix the filter widget not marking the checkbox for chosen values

attekei committed:
Fix viz settings that were not using merged remappings

iethree committed:
Remove external dimensions when FK special_type is removed

This checks field updates to sniff out when a field is going from an
FK field to something else. If it's no longer an FK field, any
external remapping should be removed as well.

mazameli commented:
So I have a MySQL table, `Games`, which has an FK field, `Pitcher ID`, which I tried to map to the `Players` table's `Last Name` field. When I do a raw data query on Games now, I get this error:

`
No implementation of method: :formatted of protocol: #'metabase.driver.generic-sql.query-processor/IGenericSQLFormattable found for class: metabase.query_processor.interface.FieldPlaceholder
`

mazameli commented:
There are a lot of arrows going on in this part of the form. I wonder if it would read better if the chevron between the two drop downs was replaced by a string like, "which maps to".

![screen shot 2017-07-12 at 10 50 00 am](https://user-images.githubusercontent.com/2223916/28131989-6af10fe0-66f0-11e7-84ed-f9edec649e86.png)

camsaul commented:
@mazameli @senior it sounds like the `FieldPlaceholder` in question isn't getting fetched and replaced with a `Field` which is what needs to happen for the SQL driver code to know how to turn it into SQL

Not sure if that's a new bug or not

senior commented:
@camsaul @mazameli I'll get it fixed up, did you get a stacktrace with the error message?

camsaul commented:
@senior in `resolve.clj` there's a protocol called IResolve that's supposed to do a couple things like tell you if something has any unresolved fields. Maybe some new clause just needs an implementation for that?

mazameli commented:
This is not a regression and not the fault of this branch, but [I still wish I could get the distribution of an FK field](https://github.com/metabase/metabase/issues/4913#issuecomment-299545296).

![screen shot 2017-07-12 at 11 18 00 am](https://user-images.githubusercontent.com/2223916/28133126-cb055cc6-66f3-11e7-866d-8e8d81b1b4cb.png)

mazameli commented:
When sorting by a field that's had its ID replaced with a string, I'd expect the field to be sorted alphabetically, not by the underlying number:

![screen shot 2017-07-12 at 11 24 24 am](https://user-images.githubusercontent.com/2223916/28133383-ad7e11f6-66f4-11e7-96e7-aa249af73162.png)

mazameli commented:
This is also related to an already existing issue to make it so that FK/numeric category fields can't be summed, averaged, and probably not minned or maxed, but these options look especially dumb when these numeric IDs have been replaced with strings:

![screen shot 2017-07-12 at 11 29 03 am](https://user-images.githubusercontent.com/2223916/28133572-55198490-66f5-11e7-8f21-382a962ac24a.png)

attekei committed:
Hide non-remapped x-axis values, disable drag-to-filter for remapped strings

kdoh committed:
only change trigger styles

kdoh committed:
Merge branch 'custom-user-projections' of github.com:metabase/metabase into custom-user-projections

attekei committed:
Show remapped names in multiscalar legends

attekei committed:
Fix problem with single value category filter not shown correctly in FilterWidget

attekei committed:
Reload metadata always when showing FieldApp

mazameli commented:
If you map a dashboard filter to a remapped field, the dropdowns get really confused:

![screen shot 2017-07-12 at 1 48 20 pm](https://user-images.githubusercontent.com/2223916/28139373-e529b5ec-6708-11e7-8adf-45af815a285b.png)

![screen shot 2017-07-12 at 1 48 47 pm](https://user-images.githubusercontent.com/2223916/28139376-e8b0c3a4-6708-11e7-8b40-813a90c01ac4.png)

![screen shot 2017-07-12 at 1 48 59 pm](https://user-images.githubusercontent.com/2223916/28139380-eac8d6a4-6708-11e7-9185-3e2aec5cd083.png)

attekei commented:
I start collecting issues found in dashboards, embedding/sharing and pulses to this comment.

- Not sure if this is considered serious or not, but for remappings from number to string we shouldn't be showing "Equal" or "Not Equal" filter drill-through choices
![screen shot 2017-07-12 at 13 50 32](https://user-images.githubusercontent.com/6034928/28139491-41810ee4-6709-11e7-81f3-f3c5f6f70294.png)

mazameli commented:
This is how a custom remapped field shows up in an emailed pulse. It basically duplicates the field, displaying both the ID as well as the remapped field. Probably not exactly what we're intending, but doesn't seem awful.

![screen shot 2017-07-12 at 2 12 09 pm](https://user-images.githubusercontent.com/2223916/28140196-2efcd368-670c-11e7-9cc1-c29bf5e853ab.png)

mazameli commented:
This one's kinda wacky: I remapped all the values of Rating in the Reviews sample table to be strings, but when doing a count of rows by Rating, the x-axis continues to display a continuous numeric axis, interrupted by my string values. I replaced _all_ the present values in Ratings (e.g., `1.7` => `Bad`), not just the integer values, FYI.

![screen shot 2017-07-12 at 2 12 33 pm](https://user-images.githubusercontent.com/2223916/28140240-6367c4f0-670c-11e7-9762-d62599df203c.png)

attekei committed:
Fix category widget, part 1

attekei committed:
A really awkward fix to a weird field value property

attekei committed:
Add tests for filter adding/removal and table viz settings

attekei committed:
Mark filter tests as pending until fixture db has Category field values

attekei committed:
Fix default FK remapping in field app tests, remove `fit`s

attekei committed:
Comment out spammy deprecated field value comments

iethree committed:
Fix bug with fk_target_field_id getting removed

Specifically updating a field attribute that wasn't it's type or it's
fk_target_field_id would cause this issue

attekei committed:
Fix code style and flow types, run full test suite [ci e2e]

iethree committed:
Field update should have coerced special_type to a keyword

attekei committed:
Proper save statuses, fix cases where dimension doesn't get reseted

attekei committed:
Improve 'should reset FK' logic in admin settings

attekei committed:
Include dimensions to PUT field/:id response

attekei committed:
Fix integrated tests [ci e2e]

iethree committed:
Add remapping awareness to pulses

iethree committed:
Revert "Remove the category type predicate from remapping"

This reverts commit 95a65d4239c4081728e12fc184f1a464ca5c1025.

iethree committed:
Apply field-should-have-field-values? predicate to query_metadata

This makes the query_metadata route behavior similar to the field
values route.

attekei committed:
Always show date widget for dates in dashboards

attekei committed:
Rename WIDGETS -> DATE_WIDGETS

attekei commented:
Now dashboards always show date picker widgets for date fields. Couldn't find other frontend issues so after adding the enum special type and doing the remaining backend changes this should be ready to be merged.

attekei committed:
Merge branch 'master' of https://github.com/metabase/metabase into custom-user-projections

attekei committed:
Use metabase-lib in FieldApp

attekei committed:
Remove over-engineered logic that resets FK remapping on FK target field change

attekei committed:
Run integrated tests [ci e2e]

attekei committed:
Fix lint and flow errors [ci e2e]

senior commented:
@mazameli I think I'll need to pair with you on the protocol error you found. My guess is there is some node in the query that isn't getting resolved properly (as @camsaul suggests) which means the code that walks that plan probably isn't getting down to that node for some reason. Getting the plan before and after it's been resolved is probably enough to figure out what's going on.

I hooked up to the retrosheet database that stats uses and was not able to reproduce the problem. I setup an external remapping for the winning pitcher and the save pitcher, both worked as expected.

attekei committed:
Fix pending editor bar test by using Rating field [ci e2e]

iethree committed:
Merge remote-tracking branch 'upstream/master' into custom-user-projections

iethree committed:
Ensure the category_id special type is restored after changed

One of the tests added recently changes the special_type of
category_id. It should have save what it's original value was and
restored it after it was done instead of assuming it's previous value
was nil.

iethree committed:
Add a new enum type for remapping

Currently this is used similarly to category. As we get more remapping
features (booleans, strings etc) this will be different from
categories and won't have the existing category baggage.

iethree committed:
Remove debugging cruft

iethree committed:
Add a validation of dimension POST for human_readable_field_id

If we have an external remapping, human_readable_field_id is
required. We now return a 400 if that is missing.

attekei committed:
Show popover and always select a field when choosing 'Use foreign key' [ci e2e]

attekei committed:
Use entity name by default, otherwise ask user to choose the field

attekei committed:
Add the Enum type to core.js

iethree committed:
Add support for sorting by an externally remapped column

When an MBQL query has an order by for a column that has been
remapped via FK, it will swap the order by for the original field for
the remapped field. Previously this would order by the original source
column and apply that to the remapping which would be unexpected.

iethree committed:
Add permissions checks for dimension updates

attekei committed:
Use fixture for reseting created at field, fix reset test failure

attekei committed:
Delete unintentional `store.dispatch`

iethree committed:
Fix bug with remapping but no order by

When ordery-by was nil, it was using in a mapv intended to replace
remapped values with what they have been remapped to. This cause
failures in the query_processor as nil was expected there, not an
empty list.

iethree committed:
Add a test for field values and sensitive visibility types

kdoh committed:
error message styling

kdoh committed:
Merge branch 'custom-user-projections' of github.com:metabase/metabase into custom-user-projections

kdoh committed:
add breadcrumbs

kdoh committed:
add tip about field name

kdoh committed:
add breadcrumb links

iethree committed:
Fixup order-by test failure

The previous commit fixed a bug when there is no order by, but there
is a remapping. The corresponding change to the
add-dimmension-projections-test didn't happen. This fixes that test failure.

iethree committed:
Clear dimension data when the special_type is not supported

If a field is set to an enum or category type and has a remapping,
that should be cleared if the field moves to a special_type that is
not allowed to have remappings (i.e. text). This commit adds checks
for that and clears the dimension in that case.

iethree committed:
Clear dimension data when the special_type is not supported

If a field is set to an enum or category type and has a remapping,
that should be cleared if the field moves to a special_type that is
not allowed to have remappings (i.e. text). This commit adds checks
for that and clears the dimension in that case.

kdoh committed:
Merge branch 'custom-user-projections' of github.com:metabase/metabase into custom-user-projections

attekei committed:
Fix integrated tests [ci all]

iethree committed:
Switch the external remapping order by test to fk engine [ci all]

External remappings are only supported on drivers that have foreign
keys. A new test was added to for ordering by the remapped but it
wasn't narrowed to only engines that support foreign keys. This commit
fixes that text.

tlrobinson commented:
This is looking great. A couple things I noticed on the admin side:

0. The select widgets feel super large
0. Do custom remapping only work for numeric fields? I see the option for Products.Rating but not Products.Category.
0. If "Use Original Value" is the only option should we just hide that section entirely?
0. I don't think "Use foreign key" is the clearest terminology. Maybe something like "Use a field from $FOREIGN_TABLE_NAME"? 
0. If the foreign table has a column marked as EntityName shouldn't we pick that as the default? Either set up the mapping on initial sync, or at least when you select "Use foreign key".
0. Also, are we planning on setting up default mappings for the Sample Dataset? That would be nifty.

kdoh committed:
Merge remote-tracking branch 'origin/master' into custom-user-projections

camsaul commented on src/metabase/api/field.clj:
```diff
@@ -65,16 +100,49 @@
+
+(api/defendpoint POST "/:id/dimension"
+  "Sets the dimension for the given field at ID"
+  [id :as {{dimension-type :type dimension-name :name human_readable_field_id :human_readable_field_id} :body}]
```
There's a caveat when doing map destructuring that renames keys for API endpoints: the API documentation generation code looks at the values in the param validation schema map and uses that to generate the endpoint description. 

Try it for yourself -- run `lein run api-documentation` -- and you'll see that the documentation will tell people they need to supply `dimension-type` (even though the actual name of the param is `type`). However avoiding renaming (e.g. a simple `:keys` destructure) would leave us with the dilemma of `name` and `type` overshadowing the `clojure.core` fns.

I think there's two ways we can have our cake and eat it too here:

1.  We could just rename these API parameters to things like `dimension_type` and `dimension_name` so the renaming during destructuring isn't needed. However this is overly verbose and I don't like naming things certain names just to make API documentation code happy
2. (Better) we can add a new wrapper function that allows you to specify an "actual" param name. Something like:

    ```clojure
    {dimension-type (su/api-key "type" (s/enum "internal" "external"))}
    ```

    Or perhaps

    ```clojure
    {(su/api-key "type" dimension-type) (s/enum "internal" "external")}
    ```

    We already have something similar with the `with-api-error-message` in the schema utils namespace, so this would largely be a matter of copying this pattern and tweaking the functions used by the `defendpoint` macro for generating documentation.
 
    This would have the added bonus of letting us do renaming destructures in API endpoints more often which would let us keep more things in `lisp-case` and make the code prettier lMO.

All that being said, this PR is already sort of huge so tweaks to the defendpoint macro would probably better be done as a follow-on so this PR doesn't get too huge. I think maybe the best option is to open an issue to remind ourselves to look at it in the future. I can do so if you want, I already have a pretty idea how to get it working nicely

senior replied:
@camsaul You writing it up sounds good, I'll work on it as a separate thing after the in-flight stuff gets done

attekei committed:
Update dimension name when changing field name, dethrottle inputs [ci all]

camsaul commented on src/metabase/api/field.clj:
```diff
@@ -86,21 +154,45 @@
+  [field value-pairs]
+  (let [human-readable-values? (validate-human-readable-pairs value-pairs)]
+    (db/insert! FieldValues
+      :field_id (:id field)
```
In cases like this we have a function in `util` called `get-id`. e.g. `(u/get-id field)`. There's a couple advantages here: 

it's guaranteed to return an integer ID, or it will throw an Exception along the lines of `not something with an ID: <whatever the arg was>`. So if `field` is somehow `nil` that would be a little easier to read than the SQL exception about violated `NOT NULL` constraints would be. Plus it would save us a failing DB call. In other places where we wouldn't be tripping a constraint it sort of serves as an assertion making sure we're actually getting a valid object, meaning you get the equivalent of a `:pre` precondition check for free

The other advantage is `get-id`is that works with either an object *or* ID: 

```clojure
(u/get-id {:id 1}) ; -> 1
(u/get-id 1)       ; -> 1
(u/get-id nil)     ; -> Exception!
```

That means you could write the function like

```clojure
(defn- create-field-values! [field-or-id value-pairs]
  ...
  (db/insert! FieldValues
    :field_id (u/get-id field-or-id)))
```

which gives you some extra flexibility.

Normally I'm weary of functions that do something we already had a perfectly fine way of doing (e.g., using `:id`) but I think the extra magic on offer justifies the increased cognitive load.

Either way no need to changes things here, I'm just letting you know it exists for future code if you see some places where using it makes more sense

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -10,14 +10,16 @@
              [db :as db]
-             [hydrate :refer [hydrate]]]))
+             [hydrate :refer [hydrate]]]
+            [metabase.query :as q]))
```
ns sorting 😉

camsaul commented on src/metabase/api/public.clj:
```diff
@@ -16,7 +16,7 @@
              [dashboard-card-series :refer [DashboardCardSeries]]
              [field-values :refer [FieldValues]]]
-            [metabase.query-processor.expand :as ql]
+            [metabase.query-processor.middleware.expand :as ql]
```
normally for a long-running PR like this if you do significant refactoring I would usually try to pull things like renaming a namespace into a separate PR and merge that ASAP, and then rebase the PR on master once it's merged. It makes it easier to see the actual changes that go on vs hundreds of "changes" that are nothing more than a renamed file. Not sure if this is the most efficient way to do it but I usually do

```bash
git checkout master
git pull
git merge <branch>
git reset origin/master
```

which will give you all the changes in the branch in an uncommited state. Then I usually go back and toss everything besides the "code cleanup"/refactoring and submit that as a PR.

(For next time)

camsaul commented on src/metabase/models/field.clj:
```diff
@@ -6,13 +6,15 @@
              [db :as db]
-             [models :as models]]))
+             [models :as models]]
+            [metabase.models.field-values :as fv]))
```
duplicate entry

camsaul commented on src/metabase/query_processor.clj:
```diff
@@ -85,10 +88,13 @@
+      add-dim/add-remapping
+      implicit-clauses/add-implicit-clauses
+      source-table/resolve-source-table-middleware
+      expand/expand-middleware                      ; ▲▲▲ QUERY EXPANSION POINT  ▲▲▲ All functions *above* will see EXPANDED query during PRE-PROCESSING
```
this is super nitpicky but can we keep the EOL comments aligned? it doesn't matter but it makes me itchy and I will literally be compelled to fix it next time I run across this namespace

camsaul commented on src/metabase/query_processor/interface.clj:
```diff
@@ -1,8 +1,10 @@
+   should go in `metabase.query-processor.middleware.expand`."
+  (:require [metabase.models
+             [field :as field]
+             [dimension :as dim]]
```
sorting

camsaul reviewed (approved):
Overall this looks really good. If you want to fix a few of the random style nitpicks like the require sorting and comment indentation I won't object. But I wouldn't consider those blockers to merging this
